lib/logstash/inputs/rabbitmq.rb

```diff
@@ -183,14 +190,30 @@ def setup!
         bind_exchange!
         @hare_info.channel.prefetch = @prefetch_count
       rescue => e
+        # when encountering an exception during shut-down,
+        # re-raise the exception instead of retrying
+        raise if stop?
+
+        reset!
+
         @logger.warn("Error while setting up connection for rabbitmq input! Will retry.",
-                     :message => e.message,
-                     :class => e.class.name,
+                     :message  => e.message,
+                     :class    => e.class.name,
                      :location => e.backtrace.first)
         sleep_for_retry
         retry
       end
 
+      # reset a partially-established connection, enabling subsequent
+      # call to `RabbitMQ#setup!` to succeed.
+      #
+      # @api private
+      def reset!
+        @hare_info.connection && @hare_info.connection.close
```

Comment:
maybe we should be prepared if `connection.close` [raises an exception](https://www.rabbitmq.com/releases/rabbitmq-java-client/v3.5.4/rabbitmq-java-client-javadoc-3.5.4/com/rabbitmq/client/Connection.html#close())

Reply:
IIUC if there is an exception in `reset!` it will bubble up from the `run` method and the plugin will be restarted so I think it is ok but OTOH this will defeat the purpose of having a specific retry loop in the `setup!` method, in that respect it might be more consistent to rescue+log+ignore any exception in `reset!` ?